### PR TITLE
Fix function declaration params comments

### LINKED ISSUE #727: Comments in function declaration parameters are moved outside parens
This seems related to https://github.com/jlongster/prettier/issues/719

Before:
```js
function foo(/* unused1, unused2 */) {}
```
After:
```js
function foo /* unused1, unused2 */() {}
```

https://jlongster.github.io/prettier/#%7B%22content%22%3A%22function%20foo(%2F*%20unused1%2C%20unused2%20*%2F)%20%7B%7D%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

This seems like unexpected behavior, but feel free to close this if it is.

### PR DESCRIPTION
**Pull request by yamafaktory**
This should fix #727.

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -1996,6 +2000,8 @@ function printFunctionParams(path, print, options) {
 
   return concat([
     "(",
+    comments.printDanglingComments(path, options, /* sameIndent */ true),
```
Comment (yamafaktory) [inline src/printer.js]: This is clearly an opinionated choice here. I try to stay consistent with the guidelines provided by @jlongster for regarding comments positioning in template literals.

File: src/printer.js [hunk 2]
```diff
@@ -2903,6 +2909,25 @@ function isFlowNodeStartingWithDeclare(node, options) {
     .match(/declare\s*$/);
 }
 
+function hasDanglingComments(node) {
```
Comment (vjeux) [inline src/printer.js]: You don't need this, you can do

```js
const dangling = printDanglingComments(...)
if (dangling) { ... }
```

File: src/printer.js (hunk 1 above)
Comment (vjeux) [inline src/printer.js]: You shouldn't have dangling comments if there are arguments, the comments are going to be attached to the elements. Have you seen a case when that's not happening?

File: src/comments.js [hunk 3]
```diff
@@ -440,6 +441,14 @@ function handleTemplateLiteralComments(enclosingNode, comment) {
   return false;
 }
 
+function handleFunctionDeclarationComments(enclosingNode, followingNode, comment) {
+  if (enclosingNode && enclosingNode.type === "FunctionDeclaration") {
```
Comment (vjeux) [inline src/comments.js]: probably want to add `!precedingNode && !followingNode`

File: tests/flow/binding/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -450,7 +450,7 @@ function type_shadow_nested_scope() {
 
 // fn params name clash
 
-function fn_params_name_clash(x, x /* error: x already bound */) {}
+function fn_params_name_clash(/* error: x already bound */ x, x) {}
```
Comment (jlongster) [inline tests/flow/binding/__snapshots__/jsfmt.spec.js.snap]: This is a regression. I'm not sure what this PR is trying to fix, although TBH I haven't read the original issue yet. The previous comment placement (after the var) is correct.

File: src/printer.js (hunk 1 above)
Comment (yamafaktory) [inline src/printer.js]: No, but the reason was to be able to add the comments when no params are present and I can't add the comments on the params node (which is empty by the way).

File: src/printer.js (hunk 2 above)
Comment (yamafaktory) [inline src/printer.js]: Yeah, you're totally right and anyway I'm going to refactor it to only use dangling comments when there are no params.

File: src/comments.js (hunk 3 above)
Comment (yamafaktory) [inline src/comments.js]: The way I refactored it now, I don't think we need it.

**Comment by yamafaktory** [comment]
@jlongster I've refactored it so that there's no more regression. This only use dangling comments in order to fix the case when a function has no arguments (no params).

Now, this is kept as before:

```js
function foo(/* unused1, unused2 */) {}
```

Without the fix, we got the following (not really nice):

```js
function foo /* unused1, unused2 */() {}
```

**Comment by jlongster** [comment]
👍 Thanks so much @yamafaktory !

**Comment by yamafaktory** [comment]
@jlongster thanks for letting me contributing to this awesome project!

**Pull request merged**